File: java/Sonicky/sonicky/src/main/java/com/egglang/sonicky/codec/Decoder.java
/*
 * Copyright 2017 egglang <t.egawa@gmail.com>.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.egglang.sonicky.codec;

import android.media.AudioFormat;
import android.media.AudioRecord;
import android.media.MediaRecorder;
import android.util.Log;

import com.egglang.sonicky.ecc.EccDecoder;
import com.egglang.sonicky.ecc.EccInstanceProvider;

import java.nio.ByteBuffer;
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.Deque;
import java.util.List;

import static com.egglang.sonicky.codec.SettingValues.BASELINE;
import static com.egglang.sonicky.codec.SettingValues.BIT_DURATION;
import static com.egglang.sonicky.codec.SettingValues.CHAR_FREQ;
import static com.egglang.sonicky.codec.SettingValues.CHAR_THRESH;
import static com.egglang.sonicky.codec.SettingValues.CHUNK_SIZE;
import static com.egglang.sonicky.codec.SettingValues.IDLE_LIMIT;
import static com.egglang.sonicky.codec.SettingValues.RATE;

public class Decoder {
    private static final int AUDIO_FORMAT = AudioFormat.ENCODING_PCM_16BIT;
    private static final int AUDIOBUF_SIZE =
            AudioRecord.getMinBufferSize(
                    RATE,
                    AudioFormat.CHANNEL_IN_MONO,
                    AUDIO_FORMAT);

    private static final double[] WINDOW;

    static {
        WINDOW = new double[CHUNK_SIZE];
        for (int i = 0; i < CHUNK_SIZE; i++) { // hamming
            WINDOW[i] = 0.54 - 0.46 * Math.cos(2.0 * Math.PI * i / (CHUNK_SIZE - 1));
        }
    }

    private int mWinLen;
    private int mWinFudge;
    private int mBufLen;
    private int mIdleCount;
    private Deque<Integer> mBuffer;
    private Deque<Integer> mBytes;
    private byte[] mReceivedBytes = new byte[]{};
    private double[] mAudio;
    private boolean mFinished;
    private AudioRecord mAudioRec = null;
    private EccDecoder mEccDecoder;

    public Decoder(boolean eccEnabled) {
        init_(eccEnabled);
    }

    private void init_(boolean eccEnabled) {
        mWinLen = (int) (BIT_DURATION * RATE / CHUNK_SIZE);
        mWinFudge = mWinLen / 2;
        mBuffer = new ArrayDeque<>();
        mBytes = new ArrayDeque<>();
        mBufLen = mWinLen + mWinFudge;
        mIdleCount = 0;
        mEccDecoder = EccInstanceProvider.getDecoder(eccEnabled);

        Log.d(getClass().getName(), String.format("AUDIOBUF_SIZE: %d", AUDIOBUF_SIZE));
        Log.d(getClass().getName(), String.format("CHUNK_SIZE: %d", CHUNK_SIZE));
        Log.d(getClass().getName(), String.format("mWinLen: %d", mWinLen));
        Log.d(getClass().getName(), String.format("mWinFudge: %d", mWinFudge));
        Log.d(getClass().getName(), String.format("mBufLen: %d", mBufLen));

        setupAudioRecord();
    }

    private void setupAudioRecord() {
        mAudio = new double[CHUNK_SIZE];

        mFinished = false;

        mAudioRec = new AudioRecord(
                MediaRecorder.AudioSource.MIC,
                RATE, AudioFormat.CHANNEL_IN_MONO,
                AUDIO_FORMAT,
                AUDIOBUF_SIZE);
        mAudioRec.setRecordPositionUpdateListener(new AudioRecord.OnRecordPositionUpdateListener() {
            @Override
            public void onMarkerReached(AudioRecord recorder) {
                //
            }

            @Override
            public void onPeriodicNotification(AudioRecord recorder) {
                short[] audioData = new short[CHUNK_SIZE];
                int size = recorder.read(audioData, 0, CHUNK_SIZE);
                window(audioData);

                double[] powerlist = new double[18];
                for (int i = 0; i < powerlist.length; i++) {
                    powerlist[i] = goertzel(CHAR_FREQ[i]);
                }
                double base = goertzel(BASELINE);

                updateState(powerlist, base);
                signalToBits();
                processByte();
            }
        });
        mAudioRec.setPositionNotificationPeriod(CHUNK_SIZE);
    }

    public byte[] listen() {
        mReceivedBytes = new byte[]{};
        mFinished = false;

        if (mAudioRec.getRecordingState() != AudioRecord.RECORDSTATE_RECORDING) {
            mAudioRec.startRecording();
        }

        while (true) {
            if (mFinished) break;

//            short[] audioData = readAudioData();
//            window(audioData);
//
//            int[] powerlist = new int[18];
//            for (int i = 0; i < powerlist.length; i++) {
//                powerlist[i] = goertzel(CHAR_FREQ[i]);
//            }
//            int base = goertzel(BASELINE);
//
//            updateState(powerlist, base);
//            signalToBits();
//            processByte();
        }
        return mReceivedBytes;
    }

    public void stop() {
        mFinished = true;
    }

    public void quit() {
        mFinished = true;
        mAudioRec.stop();
    }

    private short[] readAudioData() {
        short[] buffer = new short[CHUNK_SIZE];
        mAudioRec.read(buffer, 0, CHUNK_SIZE);
        return buffer;
    }

    private void window(short[] audioDatas) {
        for (int i = 0; i < WINDOW.length; i++) {
            double audio = (double) audioDatas[i];
            double window = WINDOW[i];
            mAudio[i] = audio * window;
        }
    }

    private double goertzel(int frequency) {
        double prev1 = 0.0;
        double prev2 = 0.0;
        double normFreq = frequency / (double) RATE;
        double coeff = 2 * Math.cos(2 * Math.PI * normFreq);
        double s;
        double power;

        for (double sample : mAudio) {
            s = sample + (coeff * prev1) - prev2;
            prev2 = prev1;
            prev1 = s;
        }
        power = (prev2 * prev2) + (prev1 * prev1) - (coeff * prev1 * prev2);
        return (power) + 1;
    }

    private void updateState(double[] powerlist, double base) {
        int state = -3; // silent
        final int arrayLength = powerlist.length;

        double[] pw = new double[powerlist.length];
        int[] judge = new int[arrayLength];

        double pwMax = 0;
        int sum = 0;

        for (int i = 0; i < arrayLength; i++) {
            pw[i] = powerlist[i] / base;
            judge[i] = pw[i] > CHAR_THRESH[i] ? 1 : 0;
            if (judge[i] == 0) {
                pw[i] = 0;
            }
            sum += judge[i];
        }

        if (sum > 0) {
            int maxIndex = 0;
            for (int i = 0; i < arrayLength; i++) {
                if (pwMax < pw[i]) {
                    pwMax = pw[i];
                    maxIndex = i;
                }
            }
            state = maxIndex - 2;
        }

        if (mBuffer.size() >= mBufLen) {
            mBuffer.pollFirst();

        }

        mBuffer.add(state);
    }

    private void signalToBits() {
        if (mBuffer.size() < mBufLen) {
            return;
        }

        List<Integer> buf = new ArrayList<>();
        for (int db : mBuffer) {
            buf.add(db);
        }

        final int SIGNALS_COUNT = CHAR_FREQ.length + 1;
        Costs costs = new Costs(SIGNALS_COUNT);

        for (int i = 0; i < mWinFudge; i++) {
            List<Integer> win = new ArrayList<Integer>(mWinLen);
            for (int j = i; j < (mWinLen + i); j++) win.add(buf.get(j));
            for (int k = 0; k < SIGNALS_COUNT; k++) {
                costs.append(k, sum(k, win));
            }
        }

        int signal = costs.getSignal();
        int fudge = costs.getFudge();
        for (int j = 0; j < (mWinLen + fudge); j++) {
            mBuffer.pollFirst();
        }

        // If we got a signal, put it in the byte!
        if (signal < 16) {
            mBytes.add(signal);
            Log.d("Signal:",
                    String.format("s: %s , data: %s, receive data: %s",
                            signal,
                            Arrays.toString(mBytes.toArray()),
                            Arrays.toString(mReceivedBytes))
            );
        }
        // If we get a charstart signal, reset byte!
        else if (signal == 16) {
            mBytes.clear();
        }
        // If we get a charend signal, reset byte and mBuffer!
        else if (signal == 17) {
            mBytes.clear();
            mBuffer.clear();
            stop();
        }
        // If we get no signal, increment idlecount if we are idling
        if (signal == 18) {
            mIdleCount++;
        } else {
            mIdleCount = 0;
        }
        if (IDLE_LIMIT <= mIdleCount) {
            mIdleCount = 0;
            mBytes.clear();
            clearReceivedBytes();
        }
    }

    private void clearReceivedBytes() {
        mReceivedBytes = new byte[]{};
    }

    private void processByte() {
        int len = mBytes.size();
        if (len != mEccDecoder.getLen()) {
            return;
        }
        Integer[] bytes = mBytes.toArray(new Integer[0]);
        int[] resultBytes = new int[len / 2];
        for (int i = 0; i < len; i += 2) {
            resultBytes[i / 2] = (bytes[i] << 4) + bytes[i + 1];
        }
        byte[] newByte = mEccDecoder.decode(resultBytes);
        ByteBuffer tmpbuf = ByteBuffer.allocate(mReceivedBytes.length + newByte.length);
        tmpbuf.put(mReceivedBytes);
        tmpbuf.put(newByte);
        mReceivedBytes = tmpbuf.array();
        mBytes.clear();
    }

    private int sum(int base, Collection<Integer> win) {
        int result = 0;
        if (base == 16) {
            base = -1;
        }
        if (base == 17) {
            base = -2;
        }
        if (base == 18) {
            base = -3;
        }
        for (Integer value : win) {
            if (value != base) {
                result += 1;
            }
        }
        return result;
    }

    static class Costs {
        List<Integer>[] costs;

        Costs(int size) {
            initialize(size);
        }

        private void initialize(int size) {
            costs = new ArrayList[size];
            for (int i = 0; i < size; i++) {
                costs[i] = new ArrayList<>();
            }
        }

        private void append(int index, int v) {
            costs[index].add(v);
        }

        private int[] minCosts() {
            int[] minCosts = new int[costs.length];
            for (int i = 0; i < costs.length; i++) {
                minCosts[i] = min(costs[i]);
            }
            return minCosts;
        }

        private int minCost() {
            int[] minCosts = minCosts();
            int minCost = Integer.MAX_VALUE;
            for (int minCost1 : minCosts) {
                if (minCost1 < minCost) {
                    minCost = minCost1;
                }
            }
            return minCost;
        }

        private int min(List<Integer> c) {
            int min = 65536;
            for (Integer i : c) {
                if (i < min) {
                    min = i;
                }
            }
            return min;
        }

        private int getFudge() {
            int minCost = minCost();
            // getSignal
            int signal = getSignal(); // FIXME redundant
            // getFudge
            return costs[signal].indexOf(minCost);
        }

        private int getSignal() {
            int[] minCosts = minCosts();
            int minCost = Integer.MAX_VALUE;
            int signal = 0;
            for (int i = 0; i < minCosts.length; i++) {
                if (minCosts[i] < minCost) {
                    minCost = minCosts[i];
                    signal = i;
                }
            }
            return signal;
        }
    }
}

File: py/sonicky/codec/decoder.py
# -*- coding: utf-8 -*-
#
# Copyright 2017 egglang <t.egawa@gmail.com>
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import division
from collections import deque
import struct
import math
import sys

import pyaudio
import numpy as np

from sonicky.codec.conf import Settings
from sonicky.codec.ecc import EmptyEcc


class Decoder:
    def __init__(self, debug=True, coder=EmptyEcc(), setting=Settings()):
        self.win_len = 2 * int(setting.bit_duration * setting.rate / setting.chunk_size / 2)
        self.win_fudge = int(self.win_len / 2)
        self.buffer = deque()
        self.buf_len = self.win_len + self.win_fudge
        self.byte = []
        self.receivebytes = bytearray()
        self.idlecount = 0
        self.coder = coder
        self.do_quit = False
        self.character_callback = None
        self.idle_callback = None
        self.debug = debug
        self.setting = setting
        self.p = pyaudio.PyAudio()
        self.stream = self.p.open(format=pyaudio.paInt16,
                                  channels=1,
                                  rate=setting.rate,
                                  input=True,
                                  frames_per_buffer=setting.audiobuf_size)

    def listen(self):
        self.do_quit = False
        self.receivebytes = []

        while not self.do_quit:
            audiostr = self.stream.read(self.setting.chunk_size)
            self.audio = list(struct.unpack("%dh" % self.setting.chunk_size, audiostr))
            self.window()
            powerlist = np.array([self.goertzel(i) for i in self.setting.char_freq])
            base = self.goertzel(self.setting.baseline)
            self.update_state(powerlist, base)
            self.signal_to_bits()
            self.process_byte()
        return self.receivebytes

    def window(self):
        self.audio = [aud * win for aud, win in zip(self.audio, self.setting.window)]

    def goertzel(self, frequency):
        prev1 = prev2 = 0.0
        norm_freq = frequency / self.setting.rate
        coeff = 2 * math.cos(2 * math.pi * norm_freq)
        for sample in self.audio:
            s = sample + (coeff * prev1) - prev2
            prev2 = prev1
            prev1 = s
        power = (prev2 * prev2) + (prev1 * prev1) - (coeff * prev1 * prev2)
        return int(power) + 1  # prevents division by zero problems

    def update_state(self, powerlist, base):
        state = -3  # silent
        pw = powerlist / base
        th = np.array(self.setting.char_threshold)
        judge = pw > th
        pw[judge == False] = 0

        # Find the maximum value
        if sum(judge) > 0:
            state = int(np.argmax(pw) - 2)

        if len(self.buffer) >= self.buf_len:
            self.buffer.popleft()
        self.buffer.append(state)

    # Takes the raw noisy samples of -1/0/1 and finds the bitstream from it
    def signal_to_bits(self):
        if len(self.buffer) < self.buf_len:
            return

        buf = list(self.buffer)

        if self.debug:
            self.printbuf(buf)

        costs = [[] for i in range(19)]  #
        for i in range(self.win_fudge):
            win = buf[i: self.win_len + i]
            #
            costs[0].append(sum(x != 0 for x in win))
            costs[1].append(sum(x != 1 for x in win))
            costs[2].append(sum(x != 2 for x in win))
            costs[3].append(sum(x != 3 for x in win))
            costs[4].append(sum(x != 4 for x in win))
            costs[5].append(sum(x != 5 for x in win))
            costs[6].append(sum(x != 6 for x in win))
            costs[7].append(sum(x != 7 for x in win))
            costs[8].append(sum(x != 8 for x in win))
            costs[9].append(sum(x != 9 for x in win))
            costs[10].append(sum(x != 10 for x in win))
            costs[11].append(sum(x != 11 for x in win))
            costs[12].append(sum(x != 12 for x in win))
            costs[13].append(sum(x != 13 for x in win))
            costs[14].append(sum(x != 14 for x in win))
            costs[15].append(sum(x != 15 for x in win))
            costs[16].append(sum(x != -1 for x in win))
            costs[17].append(sum(x != -2 for x in win))
            costs[18].append(sum(x != -3 for x in win))
            #
        min_costs = [min(costs[i]) for i in range(19)]  #
        min_cost = min(min_costs)
        signal = min_costs.index(min_cost)
        fudge = costs[signal].index(min_cost)
        for i in range(self.win_len + fudge):
            self.buffer.popleft()

        # got a signal
        if signal < 16:  #
            self.byte.append(signal)

        # got charstart signal.
        elif signal == 16:  #
            signal = 's'
            self.byte = []

        # got charend signal
        elif signal == 17:  #
            signal = 'e'
            self.byte = []
            self.buffer = deque()
            self.quit()

        # If we get no signal, increment idlecount
        if signal == 18:  #
            self.idlecount += 1
        else:
            self.idlecount = 0
        if self.idlecount > self.setting.idle_limit and self.idle_callback:
            self.idlecount = 0
            self.idle_callback()

        if self.debug:
            if signal == 18:  #
                signal = '-'
            sys.stdout.write('')
            sys.stdout.write('|{}|\n'.format(signal))
            sys.stdout.flush()

    def process_byte(self):
        if len(self.byte) != self.coder.expected_size():
            return
        byte = self.coder.get_decoded_bytes(self.byte)
        if byte is not None:
            self.receivebytes.append(byte)  #
        self.byte = []

    def quit(self):
        self.do_quit = True

    def printbuf(self, buf):
        newbuf = ['-' if x is -3 else x for x in buf]  #
        newbuf = ['s' if x is -1 else x for x in newbuf]  #
        newbuf = ['e' if x is -2 else x for x in newbuf]  #
        print(repr(newbuf).replace(', ', ' ').replace('\'', ''))

